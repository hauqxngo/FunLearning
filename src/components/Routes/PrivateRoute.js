import React, { useContext } from "react";
import { Route, Redirect } from "react-router-dom";
import UserContext from "../../UserContext";

/** "High-Order Component" (HOC) for private routes.
 *
 * This component will check
 * if there is a valid current user and only continues to the
 * route if so. If no user is present, redirects to login form.
 */

const PrivateRoute = ({ exact, path, children }) => {
  const { currentUser } = useContext(UserContext);

  if (!currentUser) {
    return <Redirect to="/login" />;
  }

  return (
    <Route exact={exact} path={path}>
      {children}
    </Route>
  );
};

export default PrivateRoute;
