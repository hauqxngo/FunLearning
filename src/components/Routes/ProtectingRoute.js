import React, { useContext } from "react";
import { Route, Redirect } from "react-router-dom";
import UserContext from "../../UserContext";

// Prevent currentUser to access routes /signup and /login
const ProtectingRoute = ({ exact, path, children }) => {
  const { currentUser } = useContext(UserContext);

  if (currentUser) {
    return <Redirect to="/" />;
  }

  return (
    <Route exact={exact} path={path}>
      {children}
    </Route>
  );
};

export default ProtectingRoute;
