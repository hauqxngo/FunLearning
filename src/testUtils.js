import React from "react";
import UserContext from "./UserContext";

const testUser = {
  username: "testuser",
  first_name: "Test",
  last_name: "User",
  email: "test@test.com",
  photo_url: null,
};

const UserProvider = ({ children, currentUser = testUser }) => (
  <UserContext.Provider value={{ currentUser }}>
    {children}
  </UserContext.Provider>
);

export { UserProvider };
