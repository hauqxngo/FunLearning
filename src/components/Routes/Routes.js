import React from "react";
import { Switch, Route, Redirect } from "react-router-dom";
import Home from "../Pages/Home";
import CategoryList from "../Categories/CategoryList";
import CategoryDetail from "../Categories/CategoryDetail";
import LoginForm from "../Forms/LoginForm";
import SignupForm from "../Forms/SignupForm";
import ProfileForm from "../Forms/ProfileForm";
import PrivateRoute from "./PrivateRoute";
import ProtectingRoute from "./ProtectingRoute";

/** Routes are wrapped by <Private> would only visible when logged in
 *
 * Invalid links will be redirected to Home page.
 */

const Routes = ({ login, signup }) => {
  return (
    <div>
      <Switch>
        <Route exact path="/">
          <Home login={login} />
        </Route>
        <ProtectingRoute exact path="/login">
          <LoginForm login={login} />
        </ProtectingRoute>
        <ProtectingRoute exact path="/signup">
          <SignupForm signup={signup} />
        </ProtectingRoute>

        <PrivateRoute exact path="/categories">
          <CategoryList />
        </PrivateRoute>
        <PrivateRoute exact path="/categories/:handle">
          <CategoryDetail />
        </PrivateRoute>
        <PrivateRoute exact path="/profile">
          <ProfileForm />
        </PrivateRoute>

        <Redirect to="/" />
      </Switch>
    </div>
  );
};

export default Routes;
