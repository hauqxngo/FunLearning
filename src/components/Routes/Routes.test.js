import React from "react";
import { render } from "@testing-library/react";
import Routes from "./Routes";
import { MemoryRouter } from "react-router";
import { UserProvider } from "../../testUtils";

it("renders without crashing", () => {
  render(
    <MemoryRouter>
      <UserProvider>
        <Routes />
      </UserProvider>
    </MemoryRouter>
  );
});

it("matches snapshot", () => {
  const { asFragment } = render(
    <MemoryRouter>
      <UserProvider>
        <Routes />
      </UserProvider>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});
