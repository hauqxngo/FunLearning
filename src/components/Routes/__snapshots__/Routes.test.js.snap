// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <header
      id="header"
    >
      <div
        class="intro"
      >
        <div
          class="overlay"
        >
          <div
            class="container"
          >
            <div
              class="row"
            >
              <div
                class="col-md-8 col-md-offset-2 intro-text"
              >
                <div
                  class="intro-center"
                >
                  <h1>
                    Let's learn something new!
                  </h1>
                  <p>
                    Help your little ones learn the basic things and get ready for preschool.
                  </p>
                  <form>
                    <button
                      class="btn btn-outline-light mt-3"
                    >
                      <i
                        class="fas fa-chalkboard-user"
                      />
                       TRY IT OUT
                    </button>
                  </form>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </header>
    <div
      id="features"
    >
      <div
        class="container"
      >
        <div
          class="col-md-10 col-md-offset-1 section-title"
        >
          <h2>
            Features
          </h2>
        </div>
        <div
          class="row"
        >
          <div
            class="col-xs-6 col-md-3"
          >
            <a
              class="text-decoration-none"
              href="/categories/alphabet"
            >
              <i
                class="fa fa-arrow-down-a-z"
              />
              <h3
                class="text-secondary"
              >
                Alphabet
              </h3>
            </a>
            <p>
              Learn ABC phonics, letters and build vocabularies.
            </p>
          </div>
          <div
            class="col-xs-6 col-md-3"
          >
            <a
              class="text-decoration-none"
              href="/categories/numbers"
            >
              <i
                class="fa fa-list-ol"
              />
              <h3
                class="text-secondary"
              >
                Numbers
              </h3>
            </a>
            <p>
              123, count with me! Learn to count from 1 to 10.
            </p>
          </div>
          <div
            class="col-xs-6 col-md-3"
          >
            <a
              class="text-decoration-none"
              href="/categories/colors"
            >
              <i
                class="fa fa-palette"
              />
              <h3
                class="text-secondary"
              >
                Colors
              </h3>
            </a>
            <p>
              Introduce primary, secondary colors, and more!
            </p>
          </div>
          <div
            class="col-xs-6 col-md-3"
          >
            <a
              class="text-decoration-none"
              href="/categories/shapes"
            >
              <i
                class="fa fa-shapes"
              />
              <h3
                class="text-secondary"
              >
                Shapes
              </h3>
            </a>
            <p>
              Learn basic shapes with examples in every day life.
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      id="about"
    >
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="col-xs-12 col-md-6"
          >
            <img
              alt="Reading Toddler"
              class="img-responsive"
              src="images/about.jpeg"
            />
          </div>
          <div
            class="col-xs-12 col-md-6"
          >
            <div
              class="about-text"
            >
              <h2>
                About
              </h2>
              <p>
                Hi there! Thanks for checking this out.
              </p>
              <p>
                I created this app for my final project at a coding bootcamp, and behind the motivation to help entertain my 2-year-old daughter.
              </p>
              <p>
                Hope you and your kiddo(s) enjoy it!
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <section
      class="colored-section"
    >
      <div
        class="container-fluid"
      >
        <h3
          class="big-heading"
        >
          Always Free. Learn More Today.
        </h3>
        <form>
          <button
            class="cta-button btn btn-outline-light"
          >
            <i
              class="fas fa-chalkboard-user"
            />
             TRY IT OUT
          </button>
        </form>
      </div>
    </section>
    <footer
      id="footer"
    >
      <div
        class="container-fluid"
      >
        <a
          href="https://www.instagram.com/hauqxngo"
          rel="noreferrer"
          target="_blank"
        >
          <i
            class="social-icon fab fa-instagram"
          />
        </a>
        <a
          href="https://www.linkedin.com/in/hauqxngo"
          rel="noreferrer"
          target="_blank"
        >
          <i
            class="social-icon fa-brands fa-linkedin"
          />
        </a>
        <a
          href="https://github.com/hauqxngo"
          rel="noreferrer"
          target="_blank"
        >
          <i
            class="social-icon fa-brands fa-github"
          />
        </a>
        <p>
          Â© Copyright 2022 
          <a
            href="https://hauqxngo.com"
            rel="noreferrer"
            target="_blank"
          >
            Hau Ngo
          </a>
        </p>
      </div>
    </footer>
  </div>
</DocumentFragment>
`;
